dig/dig.go

```diff
@@ -30,12 +30,39 @@ import (
 )
 
 var (
-	errParamType   = errors.New("graph injection must be done through a pointer or a function")
-	errReturnCount = errors.New("constructor function must return exactly one value")
-	errReturnKind  = errors.New("constructor return type must be a pointer")
-	errArgKind     = errors.New("constructor arguments must be pointers")
+	errParamType     = errors.New("graph injection must be done through a pointer or a function")
+	errReturnCount   = errors.New("constructor function must return up to two values")
+	errReturnKind    = errors.New("constructor return type must be a pointer")
+	errReturnErrKind = errors.New("second return value of constructor must be error")
+	errArgKind       = errors.New("constructor arguments must be pointers")
+
+	_typeOfError = reflect.TypeOf((*error)(nil)).Elem()
 )
 
+// Wraps a function that returns a single value to return that value and a nil
+// error.
+func addErrorReturn(c interface{}) interface{} {
```

Comment:
This is pretty neat :)

I don't actually think we need to wrap every non-error function though. We can move that logic all the way down to `value()` method on the `funcNode`

At the moment it does `err, _ := vs[1].Interface().(error)`. Instead of doing that we can check if there is a second value and if there is, then check it. Otherwise just proceed as normal with one value